onnxscript/ir/_core.py

```diff
@@ -1236,6 +1237,42 @@ def __init__(
         self._type = type
 
 
+def _check_node_safe_to_remove(
+    node: Node, to_remove: AbstractSet[Node], graph_outputs: AbstractSet[Value]
+) -> None:
+    """Check if a node is safe to remove.
+
+    1. It checks to make sure there are no users of the node that are not
+        to be removed before removing it.
+    2. It checks the node does not contribute to any graph outputs.
+
+    This check is typically O(1) assuming the number of consumers of the node is small
+
+    Args:
+        node: The node to check.
+        to_remove: A set of nodes that are to be removed.
+            This set is used to check if the node is still being used by other
+            nodes that are not to be removed.
+        graph_outputs: A set of values that are outputs of the graph.
+
+    Raises:
+        ValueError: If the node does not belong to this graph or if there are users of the node.
+        ValueError: If the node is still being used by other nodes not to be removed.
+    """
+    for output in node.outputs:
+        if output in graph_outputs:
+            raise ValueError(
+                f"Node {node} is still an output of the graph and cannot be removed when safe=True."
+            )
+        for node, _ in output.consumers():
+            if node in to_remove:
+                continue
+            raise ValueError(
+                f"Node {node} is still being used by other nodes that are not to be "
+                f"removed: {output.consumers()}"
```

Comment:
``consumers`` returns ``self._consumers.keys()``, which is an iterator. I wonder about the error message. It should write something like ``dict_keys``.

Reply:
Updated to a tuple so that it is displayed